src/template/wasm2c_atomicops.declarations.c

```diff
@@ -160,73 +105,115 @@ DEFINE_ATOMIC_STORE(i64_atomic_store8, u8, u64)
 DEFINE_ATOMIC_STORE(i64_atomic_store16, u16, u64)
 DEFINE_ATOMIC_STORE(i64_atomic_store32, u32, u64)
 
-#define DEFINE_ATOMIC_RMW(name, op, t1, t2)                          \
-  static inline t2 name(wasm_rt_memory_t* mem, u64 addr, t2 value) { \
-    MEMCHECK(mem, addr, t1);                                         \
-    ATOMIC_ALIGNMENT_CHECK(addr, t1);                                \
-    t1 wrapped = (t1)value;                                          \
-    t1 ret = atomic_##op##_##t1(&mem->data[addr], wrapped);          \
-    return (t2)ret;                                                  \
+#define DEFINE_ATOMIC_RMW(name, opname, op, t1, t2)                            \
+  static inline t2 name(wasm_rt_memory_t* mem, u64 addr, t2 value) {           \
+    MEMCHECK(mem, addr, t1);                                                   \
+    ATOMIC_ALIGNMENT_CHECK(addr, t1);                                          \
+    t1 wrapped = (t1)value;                                                    \
+    t1 ret;                                                                    \
+    wasm_rt_memcpy(&ret, &mem->data[addr], sizeof(t1));                        \
+    ret = ret op wrapped;                                                      \
+    return (t2)ret;                                                            \
```

Comment:
hmm, shouldn't this write back to the memory...?

Reply:
ooh, good catch! will fix.